numpy/lib/tests/test__datasource.py

```diff
@@ -361,3 +361,18 @@ def test_DataSourceOpen(self):
         fp = datasource.open(local_file)
         assert_(fp)
         fp.close()
+
+def test_del_attr_handling():
+    # DataSource __del__ can be called
+    # even if __init__ fails when the
+    # Exception object is caught by the
+    # caller as happens in refguide_check
+    # is_deprecated() function
+
+    ds = datasource.DataSource()
+    # simulate failed __init__ by removing key attribute
+    # produced within __init__ and expected by __del__
+    del ds._istmpdest
+    # should not raise an AttributeError if __del__
+    # gracefully handles failed __init__:
+    ds.__del__()
```

Comment:
Isn't this going to result in `__del__` being called twice and `shutil` failing? I suppose it doesn't cause the tests to fail, but is sort of messy.

Reply:
Just trying to be thorough -- I could just delete the unit test for the 0.5 line adjustment that prevents the destructor from raising an exception in the refguide worfklow. Calling the destructor when the initialization of an object fails seems to be a source of debate in language design.

As [noted here](https://stackoverflow.com/a/7501167/2942522):

> That is why in C++ and in other object-oriented languages that have been designed with exception safety in mind, the destructor is not called if an exception is thrown in the constructor of an object